### PR TITLE
Add auto parallel cost model and unittests

### PR DESCRIPTION
**Pull request by YipZLF**
### PR types

New features

### PR changes

Others

### Describe

add cost model for distributed program generated by auto parallel module
* estimate_cost(distributed_program, cluster, process_mesh, single_cost_data, batch_size): this function takes the distributed program, cluster model and other distributed settings as inputs and estimate the runtime of one iteration and static memory and peak memory consumption.

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/cost.py [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+    def __init__(self, node, node_type, id=None):
+        self.id = id
+        self.node = node
+        self.type = node_type
+        self.cost = 0
+
+    def get_cost(self):
+        return self.cost
+
+    def set_cost(self, cost):
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/cost.py]: the get or set cost can be be implemented with python "@property"

Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thank you for your advice! Updates can be seen in commit 3ee5317.

File: python/paddle/distributed/auto_parallel/cost.py [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+        self.cluster = cluster
+        self.cost_data = single_cost_data
+        self.process_mesh = process_mesh
+
+        self.rank2dim = {}
+        self.dp_group = {}
+        self.mp_group = {}
+        self.pp_send_peer = {}
+        self.pp_recv_peer = {}
+        for pp_id in range(self.process_mesh.topology[0]):
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/cost.py]: the process_mesh.topology is not necessarily 3d. Is this appropriate?

File: python/paddle/distributed/auto_parallel/cost.py [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+    """
+    # the following line is left for now, cluster model will be involved in the future
+    assert cluster is None, "For now, cluster remains None"
+
+    cost = Cost()
+    cm_ctx = CostModelContext(
+        cluster=cluster,
+        batch_size=batch_size,
+        single_cost_data=single_cost_data,
+        process_mesh=process_mesh)
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/cost.py]: Please use CostModel to replace CostModelContext to encapsulate all functions. Besides, each op or tensor may have a different process_meshes. Please rename single_cost_data to serial_cost_data or standalone_cost_data.

File: python/paddle/distributed/auto_parallel/cost.py [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+    cm_ctx.parse_program(distributed_program)
+    cm_ctx.build_op_graph()
+    for sub_idx in range(len(distributed_program)):
+        cm_ctx.elim_multi_edges(cm_ctx.op_graph[sub_idx])
+    cm_ctx.build_rt_graph()
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/cost.py]: It's better to rename build_rt_graph to build_runtime_graph.

File: python/paddle/distributed/auto_parallel/cost.py [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+    cm_ctx.merge_comm()
+    while True:
+        cnt = 0
+        cnt += cm_ctx.merge_linear()
+        cnt += cm_ctx.merge_branch()
+        if cnt == 0:  # can't be further merged
+            break
+
+    cost.runtime = cm_ctx.get_rt_cost()
+    cost.runtime = cm_ctx.get_pipeline_time()
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/cost.py]: This estimate_cost function can be refactored: 1) construct CostModel object based on the necessary parameters and the selected mode, e.g. cost_model ;  2) call cur_cost = cost_model.get_cost() to give the results; 3) return cur_cost.

File: python/paddle/distributed/auto_parallel/cost.py [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+                succ_nodes_id = self._find_succ_op(op_id, sub_idx)
+
+                self.op_graph[sub_idx][op_id][SUCC] = succ_nodes_id
+                for succ_id in succ_nodes_id:
+                    self.op_graph[sub_idx][succ_id][PRED].append(op_id)
+
+    def build_rt_graph(self):
+        self.rt_graph = copy.deepcopy(self.op_graph)
+
+    def elim_multi_edges(self, graph=None):
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/cost.py]: It's better to rename elim to eliminate, rt to runtime, sim to simulate

File: python/paddle/distributed/auto_parallel/cost.py [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import numpy as np
+import json
+import queue
+import copy
+from enum import Enum
+import paddle
+
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/cost.py]: Please rename NodeType, NodeBase, MergedNode, CommNode, VarNode, CompNode to CostNodeType, CostNode, MergedOpsCostNode, CommOpCostNode, TensorCostNode, CompOpCostNode respectively.

**Comment by aoyulong** [comment]
Please rename cost.py to cost_model.py.

File: python/paddle/distributed/auto_parallel/cost.py [hunk 8]
```diff
@@ -0,0 +1,752 @@
...
+        return p
+
+
+class CompNode(NodeBase):
+    def __init__(self, node, node_type, id=None):
+        super(CompNode, self).__init__(node, node_type, id)
+
+    def init_comp_cost(self, cost_data):
+        op_name = self.node.type
+        if op_name in cost_data.keys():
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/cost.py]: The op_name means the op type or what? If op type，a main_program may has many ops with the same op_type,  the cost of those ops is the same?

File: python/paddle/distributed/auto_parallel/cost.py [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+                if pred.type == NodeType.VARIABLE:
+                    sim_graph[pred_id][SUCC].remove(node_id)
+                    if len(sim_graph[pred_id][
+                            SUCC]) == 0 and not pred.node.persistable:
+                        cur_mem -= pred.get_size()
+
+        return static_mem, cur_mem, top_mem
+
+    def get_pipeline_time(self):
+        if len(self.time_list) <= 1:
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/cost.py]: add check when len(self.time_list) == 0?

File: python/paddle/distributed/auto_parallel/cost.py (hunk 2 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thank you. Fixed in commit 1a66035.

File: python/paddle/distributed/auto_parallel/cost.py (hunk 3 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thanks. Fixed in commit 1a66035

File: python/paddle/distributed/auto_parallel/cost.py (hunk 4 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thanks. Fixed in commit 1a66035.

File: python/paddle/distributed/auto_parallel/cost.py (hunk 6 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thanks. Fixed in commit 1a66035.

File: python/paddle/distributed/auto_parallel/cost.py (hunk 7 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thanks. Fixed in commit 1a66035.

File: python/paddle/distributed/auto_parallel/cost.py (hunk 9 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thank you. See commit 1a66035. This value being zero indicates that the program is empty. This case is covered in the current implementation.

File: python/paddle/distributed/auto_parallel/cost.py (hunk 8 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: `op_name` serves as the key mapping to a specific op. Current implementation of cost_data uses op.type as the key. We're to improve the process of collecting cost_data in the future.

File: python/paddle/distributed/auto_parallel/cost.py (hunk 5 above)
Comment (YipZLF) [reply python/paddle/distributed/auto_parallel/cost.py]: Thank you. Fixed in commit 3ba2710

**Review by Caozhou1995 — APPROVED** [review]

File: python/paddle/fluid/tests/unittests/test_auto_parallel_cost_model.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    dist_startup_prog = []
+    for rank_id in range(NUM_RANKS):
+        partitioner = Partitioner(dist_strategy, dist_context, rank_id)
+        # logical partition
+        auto_parallel_main_prog, auto_parallel_startup_prog = partitioner.transpile_forward(
+            complete_train_program, startup_program)
+        dist_params_grads = partitioner.apply_backward(
+            loss, complete_train_program, startup_program,
+            auto_parallel_main_prog, auto_parallel_startup_prog)
+        optimizer = paddle.fluid.optimizer.AdamOptimizer()
```
Comment (fuyinno4) [inline python/paddle/fluid/tests/unittests/test_auto_parallel_cost_model.py:159]: paddle.optimizer.AdamOptimizer()

Comment (YipZLF) [reply python/paddle/fluid/tests/unittests/test_auto_parallel_cost_model.py:159]: Thanks. Will be fixed in the coming commit very soon.

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**